### PR TITLE
Export to Codesandbox

### PR DESCRIPTION
**Pull request by PeterDraex**
Storybook plugin for Export to CodeSandbox

### CODE REVIEW

File: storybook-addon-export-to-codesandbox/.github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+name: Release
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/.github/workflows/release.yml:1]: Can be this pipeline moved to a separate PR?

File: storybook-addon-export-to-codesandbox/.gitignore [hunk 2]
```diff
@@ -0,0 +1,6 @@
+dist/
+node_modules/
+storybook-static/
+build-storybook.log
+.DS_Store
+.env
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/.gitignore:6]:
```suggestion

```

nit: Not sure that this repo contains `.env` files

File: storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+module.exports = function (babel) {
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js:1]: As I remember @ecraig12345 provided a comment somewhere with JSDoc TS comments. Can you please integrate that suggestion?

File: storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+            ),
+          );
+          const expressionStatement = path.findParent(p => p.isExpressionStatement());
+          expressionStatement.insertAfter(expression);
+          path.stop();
+        }
+      },
+    },
+  };
+};
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js:26]: It will be good to add unit tests in a separate PR for this plugin via https://github.com/babel-utils/babel-plugin-tester.

File: storybook-addon-export-to-codesandbox/.storybook/main.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+// This Babel plugin adds context.parameters.fullSource property to Storybook stories,
+// which contains source of of the file where story is present.
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/.storybook/main.js:2]: It looks that this comment should in `.storybook/babel.plugin.js`

File: storybook-addon-export-to-codesandbox/LICENSE [hunk 6]
```diff
@@ -0,0 +1,21 @@
+MIT License
+
+Copyright (c) 2021 Storybook contributors
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/LICENSE:3]: This should updated I guess

File: storybook-addon-export-to-codesandbox/README.md [hunk 7]
```diff
@@ -0,0 +1,51 @@
+# Export to CodeSandbox
+
+This Storybook plugin adds "Open in CodeSandbox" button to each story displayed in Docs mode.
+
+## Installation
+
+1. Install the plugin `npm i storybook-addon-export-to-codesandbox`.
+2. Register the plugin in `.storybook/main.js` - Add `'storybook-addon-export-to-codesandbox'` to the list of addons.
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/README.md:8]: May be I am not familiar with Storybook, but where I should add it?

File: storybook-addon-export-to-codesandbox/README.md [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+3. Use .storybook/babel.plugin.js in your Storybook.
+4. Define required parameters in your `.storybook/preview.ts`:
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/README.md:10]:
```suggestion
4. Define required parameters in your `.storybook/preview.js`:
```

`preview.js` or `preview.ts`?

File: storybook-addon-export-to-codesandbox/README.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
+# Export to CodeSandbox
+
+This Storybook plugin adds "Open in CodeSandbox" button to each story displayed in Docs mode.
+
+## Installation
+
+1. Install the plugin `npm i storybook-addon-export-to-codesandbox`.
+2. Register the plugin in `.storybook/main.js` - Add `'storybook-addon-export-to-codesandbox'` to the list of addons.
+3. Use .storybook/babel.plugin.js in your Storybook.
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/README.md:9]: What I should do as a customer? Copy it?

File: storybook-addon-export-to-codesandbox/prettier.config.js [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  arrowParens: 'avoid',
+  overrides: [
+    {
+      files: 'apps/vr-tests/**/*',
+      options: {
+        // The smaller printWidth for the screener tests promotes readability by preventing test cases
+        // from being squished onto one line (and squished up against each other in consecutive lines)
+        printWidth: 100,
+      },
+    },
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/prettier.config.js:21]:
```suggestion
```

Is it relevant for this repo?

File: storybook-addon-export-to-codesandbox/prettier.config.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    {
+      // These files may be run as-is in IE 11 and must not have ES5-incompatible trailing commas
+      files: ['*.html', '*.htm'],
+      options: {
+        trailingComma: 'es5',
+      },
+    },
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/prettier.config.js:28]:
```suggestion
```

Is it relevant for this repo?

File: storybook-addon-export-to-codesandbox/scripts/eject-typescript.mjs [hunk 12]
```diff
@@ -0,0 +1,30 @@
+#!/usr/bin/env zx
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/scripts/eject-typescript.mjs:1]: I don' think that we need this script. Is it intentionally there? Can we remove it?

File: storybook-addon-export-to-codesandbox/scripts/prepublish-checks.mjs [hunk 13]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env zx
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/scripts/prepublish-checks.mjs:1]: I don' think that we need this script. Is it intentionally there? Can we remove it?

File: storybook-addon-export-to-codesandbox/tsconfig.json [hunk 14]
```diff
@@ -0,0 +1,20 @@
+{
+  "compilerOptions": {
+    "allowSyntheticDefaultImports": true,
+    "baseUrl": ".",
+    "emitDecoratorMetadata": true,
+    "esModuleInterop": true,
+    "experimentalDecorators": true,
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/tsconfig.json:7]:
```suggestion
```

Do we need it there?

File: storybook-addon-export-to-codesandbox/tsconfig.json [hunk 15]
```diff
@@ -0,0 +1,20 @@
+{
+  "compilerOptions": {
+    "allowSyntheticDefaultImports": true,
+    "baseUrl": ".",
+    "emitDecoratorMetadata": true,
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/tsconfig.json:5]:
```suggestion
```

Do we need it there?

File: storybook-addon-export-to-codesandbox/src/preset/preview.ts [hunk 16]
```diff
@@ -0,0 +1,13 @@
+/**
+ * A decorator is a way to wrap a story in extra “rendering” functionality. Many addons define decorators
+ * in order to augment stories:
+ * - with extra rendering
+ * - gather details about how a story is rendered
+ *
+ * When writing stories, decorators are typically used to wrap stories with extra markup or context mocking.
+ *
+ * https://storybook.js.org/docs/react/writing-stories/decorators#gatsby-focus-wrapper
+ */
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/src/preset/preview.ts:10]:
```suggestion
```

Do we need this comment?

File: storybook-addon-export-to-codesandbox/src/withCodeSandboxButton.ts [hunk 17]
```diff
@@ -0,0 +1,139 @@
+import { StoryFn as StoryFunction, StoryContext, useEffect, StoryWrapper } from '@storybook/addons';
+import { getParameters } from 'codesandbox-import-utils/lib/api/define';
+import dedent from 'dedent';
+
+const dependencyRegex = / from '.*?'; \/\/ codesandbox-dependency: (.*?) (.*)/g;
+const dependencySubs = " from '$1';";
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/src/withCodeSandboxButton.ts:6]: _nit:_ Can we use UPPER CASE for constants?

File: storybook-addon-export-to-codesandbox/src/withCodeSandboxButton.ts [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+
+  // extract dependencies from imports
+  const matches = replaceRelativeImports(fileContent).matchAll(/import .* from ['"](.*?)['"];/g);
+
+  for (const match of matches) {
+    if (!match[1].startsWith('react/')) {
+      const dependency = match[1];
+
+      if (!dependencies.hasOwnProperty(dependency)) {
+        if (dependency.startsWith('@fluentui/react-')) dependencies[dependency] = '^9.0.0-alpha';
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/src/withCodeSandboxButton.ts:33]:
```suggestion
        if (dependency.startsWith('@fluentui/react-')) {
          dependencies[dependency] = '^9.0.0-alpha';
        }
```

_nit_

File: storybook-addon-export-to-codesandbox/src/withCodeSandboxButton.ts [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+  let storyFile = context.parameters?.fullSource;
+
+  if (!storyFile) {
+    console.error(
+      `Export to CodeSandbox: Couldn’t find source for story ${context.story}. Did you install the babel plugin?`,
+    );
+    return false;
+  }
+
+  const requiredDependencies: { [dependencyName: string]: string } =
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/src/withCodeSandboxButton.ts:79]:
```suggestion
  const requiredDependencies: PackageDependencies =
```

This looks like a duplicated type (it's also used as a type for param), is these sense to extract it?

File: storybook-addon-export-to-codesandbox/src/withCodeSandboxButton.ts [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+  if (context.viewMode === 'docs') {
+    useEffect(() => {
+      displayToolState(`#anchor--${context.id} .docs-story`, context);
+    });
+  }
+
+  return StoryFn(context);
+};
+
+const getDependencies = (fileContent: string, dependencies: { [dependencyName: string]: string }) => {
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/src/withCodeSandboxButton.ts:18]:
```suggestion
const getDependencies = (fileContent: string, requiredDependencies: { [dependencyName: string]: string }) => {
```

_nit:_ let's use the same variable name as in its call

---

This change will highlight that this function is not pure as it mutates its input, let's clone its input to make it safer:

```tsx
const getDependencies = (fileContent: string, requiredDependencies: { [dependencyName: string]: string }) => {
  const dependencies = { ...requiredDependencies };
```

File: storybook-addon-export-to-codesandbox/src/withCodeSandboxButton.ts [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+  }
+
+  const dependencies = getDependencies(storyFile, requiredDependencies);
+  storyFile = replaceRelativeImports(storyFile);
+
+  if (storyFile.match(/import .* from ['"]\./g)) {
+    console.error(
+      dedent`Export to CodeSandbox: Story "${context.story}" contains relative import without defined package.
+             Please add the following comment to the end of each line with relative import:
+             // codesandbox-dependency: [package-name] [pacakge version]`,
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/src/withCodeSandboxButton.ts:94]:
```suggestion
             // codesandbox-dependency: [package-name] [package-version]`,
```

File: storybook-addon-export-to-codesandbox/package.json [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+  "name": "storybook-addon-export-to-codesandbox",
+  "version": "0.4.0",
+  "description": "Allows to export Story code to CodeSandbox",
+  "keywords": [
+    "storybook-addons",
+    "code"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/PeterDraex/storybook-addon-export-to-codesandbox"
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/package.json:11]: Oops, does not seem that this URL is correct

File: storybook-addon-export-to-codesandbox/prettier.config.js [hunk 23]
```diff
@@ -0,0 +1,30 @@
+// @ts-check
```
Comment (layershifter) [inline storybook-addon-export-to-codesandbox/prettier.config.js:1]: IMO this file should be moved to root

File: storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js (hunk 4 above)
Comment (PeterDraex) [reply storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js:39]: I’ve created an issue #2

File: storybook-addon-export-to-codesandbox/README.md (hunk 9 above)
Comment (PeterDraex) [reply storybook-addon-export-to-codesandbox/README.md:9]: Yes, until it’s extracted to a separate package as part of #2

File: storybook-addon-export-to-codesandbox/README.md (hunk 8 above)
Comment (PeterDraex) [reply storybook-addon-export-to-codesandbox/README.md:10]: Can be either - depends on the setup. But I’ll change it to `js`, as it’s more common I think.

File: storybook-addon-export-to-codesandbox/README.md (hunk 7 above)
Comment (PeterDraex) [reply storybook-addon-export-to-codesandbox/README.md:8]: `.storybook/main.js` is the main configuration file for storybook. That’s how they refer to it in [Storybook documentation](https://storybook.js.org/docs/react/addons/install-addons#using-addons). I’ve added an example to make it clearer.

File: storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js [hunk 24]
```diff
@@ -0,0 +1,39 @@
+/**
+ * This Babel plugin adds `context.parameters.fullSource` property to Storybook stories,
```
Comment (ling1726) [inline storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js:2]: is it possible to find any github/docs resources to what this is for ppl who don't know the internals of storybook ?

File: storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js [hunk 25]
```diff
@@ -0,0 +1,39 @@
...
+        ) {
+          const storyName = path.node.object.name;
+          const expression = t.expressionStatement(
+            t.assignmentExpression(
+              '=',
+              t.memberExpression(
+                t.memberExpression(t.identifier(storyName), t.identifier('parameters')),
+                t.identifier('fullSource'),
+              ),
+              t.identifier('__STORY__'),
```
Comment (ling1726) [inline storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js:29]: Is it possible to link to some github/docs as to what this magical `__Story__` var is ?

File: storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js (hunk 24 above)
Comment (ling1726) [inline storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js:2]: Code example of what you want to achieve would be great here, since it's a simple modification

File: storybook-addon-export-to-codesandbox/README.md (hunk 9 above)
Comment (ling1726) [reply storybook-addon-export-to-codesandbox/README.md:18]: Is it possible to have an export from the addon package with the babel plugin in the meantime ?

File: storybook-addon-export-to-codesandbox/README.md [hunk 26]
```diff
@@ -0,0 +1,60 @@
...
+};
+```
+
+## Recommended story setup
+
+Each story should be put into its own file with a `.stories.tsx` extension. This file should not contain a default export, only a single named export, which is the story. All story files should then be re-exported from a file `.stories.tsx` with a default export. See [RFC: Authoring storybook stories](https://github.com/microsoft/fluentui/blob/master/rfcs/convergence/authoring-stories.md) for more details and examples.
+
+This practice is recommended so that the "Open in CodeSandbox" button would export a single story.
+
+## Relative imports
```
Comment (ling1726) [inline storybook-addon-export-to-codesandbox/README.md:54]: FYI @ecraig12345 far better interim workaround than `ts-ignore`

File: storybook-addon-export-to-codesandbox/README.md [hunk 27]
```diff
@@ -0,0 +1,60 @@
+# Export to CodeSandbox
```
Comment (ling1726) [inline storybook-addon-export-to-codesandbox/README.md:1]: Can we also have a `CONTRIBUTING.md` to make it easier for ppl maintain this addon ?

File: storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js (hunk 25 above)
Comment (PeterDraex) [reply storybook-addon-export-to-codesandbox/.storybook/babel.plugin.js:38]: added to the top of the file

File: storybook-addon-export-to-codesandbox/package.json [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+  "name": "storybook-addon-export-to-codesandbox",
+  "version": "0.4.0",
+  "description": "Allows to export Story code to CodeSandbox",
+  "keywords": [
+    "storybook-addons",
+    "code"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/microsoft/fluentui-storybook-addons"
```
Comment (ling1726) [inline storybook-addon-export-to-codesandbox/package.json:11]: nit: I think you should add the folder to the URL since npm will pull the README from there ?

@layershifter  am I right on that ?

**Review by layershifter — APPROVED** [review]
This PR contains both `yarn.lock` & `package-lock.json`. I guess that one should be deleted, right?

File: storybook-addon-export-to-codesandbox/package.json (hunk 28 above)
Comment (layershifter) [reply storybook-addon-export-to-codesandbox/package.json:11]: Yeah, it should be nice as it is used by NPM

image

Comment (ecraig12345) [reply storybook-addon-export-to-codesandbox/package.json:11]: npm pulls the readme from the package itself, not the git repo.

From this page it looks like `repository.url` is supposed to be for the repo root. `homepage` is for the website (or readme) URL and `repository.directory` is for the subfolder. https://docs.npmjs.com/cli/v6/configuring-npm/package-json#repository

Comment (ling1726) [reply storybook-addon-export-to-codesandbox/package.json:11]: Ah thanks @ecraig12345 let's add `repository.directory` then

**Review by ling1726 — APPROVED** [review]

**Review by ling1726 — APPROVED** [review]
